src/main/scala/is/hail/variant/MatrixTable.scala

```diff
@@ -2532,4 +2535,165 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
+      entryType = matrixType.entryType ++ TStruct("prev_entries" -> TArray(matrixType.entryType))
+    )
+
+    val partitioner = rvd.partitioner
+    val adjustments = partitioner
+      .rangeBounds
+      .zipWithIndex
+      .map { case (bound, i) =>
+        val startLocus = bound.start.asInstanceOf[Row].getAs[Locus](0)
+        val minPartitionNeeded = partitioner.getPartitionPK(
+          Row(startLocus.copy(position = startLocus.position - basePairs)))
+        (minPartitionNeeded to i)
+          .map(Adjustment[RegionValue](_, identity))
+          .toArray
+      }.toArray
+
+    val localRangeTree = partitioner.rangeTree
+
+    val adjRDD = rvd.crdd.adjustPartitions(adjustments)
+
+    val localRVRowType = rvRowType
+    val locusIndex = localRVRowType.fieldIdx("locus")
+    val entriesIndex = localRVRowType.fieldIdx(MatrixType.entriesIdentifier)
+    val entryArrayType = matrixType.entryArrayType
+    val entryType = matrixType.entryType
+    val rg = this.referenceGenome
+
+    val lociStarts = sparkContext.broadcast(partitioner.rangeBounds.map { interval =>
+      val locus = interval.start.asInstanceOf[Row].getAs[Locus](0)
+    locus.copy(position = locus.position - basePairs)
+    })
+
+    val nCols = numCols
+    val newRDD = adjRDD.cmapPartitionsAndContextWithIndex { (partitionIndex, context, it) =>
+      val newContext = context.freshContext
+
+      val rb = new mutable.ArrayStack[Region]()
+      def fetchRegion(): Region = {
+        if (rb.isEmpty)
+          newContext.freshRegion
+        else
+          rb.pop()
+      }
+
+      def recycleRegion(r: Region): Unit = {
+        r.clear()
+        rb.push(r)
+      }
+      val deque = new java.util.ArrayDeque[(Locus, RegionValue)]()
+      val stack = new java.util.Stack[RegionValue]()
+
+      it.flatMap { f =>
+        val skip = it.hasNext // these rows are loaded from previous partitions
+        val rvIterator = f(newContext)
+
+        val region = context.region
+        val rv2 = RegionValue()
+        val rvb = new RegionValueBuilder()
+        rvIterator.flatMap { rv =>
+          val locus = UnsafeRow.readLocus(rv.region, localRVRowType.loadField(rv, locusIndex), rg)
+          val pos = locus.position
+
+          val result = if (skip) {
+            None
+          } else {
+            def discard(x: (Locus, RegionValue)): Boolean = x != null && (x._1.contig != locus.contig ||
+              x._1.position < locus.position - basePairs)
+
+            while (discard(deque.peekLast()))
+              recycleRegion(deque.removeLast()._2.region)
+
+            val rvs = deque.iterator().asScala.map(_._2).toArray
+
+            rvb.set(region)
+            rvb.clear()
+            region.clear()
+            rvb.start(newType.rvRowType)
+            rvb.startStruct()
+            var rfIndex = 0
+            while (rfIndex < localRVRowType.size) {
+              if (rfIndex != entriesIndex)
```

Comment:
since entriesIndex is constant, you could optimize this a bit with two while loops, same below. your call